core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1483,12 +1480,14 @@ class Log(@volatile private var _dir: File,
           s"but we only have log segments in the range $logStartOffset to $endOffset.")
 
       val maxOffsetMetadata = isolation match {
-        case FetchLogEnd => nextOffsetMetadata
+        case FetchLogEnd => endOffsetMetadata
         case FetchHighWatermark => fetchHighWatermarkMetadata
         case FetchTxnCommitted => fetchLastStableOffsetMetadata
       }
 
-      if (startOffset > maxOffsetMetadata.messageOffset) {
+      if (startOffset == maxOffsetMetadata.messageOffset) {
+        return emptyFetchDataInfo(maxOffsetMetadata, includeAbortedTxns)
```

Comment:
If the start offset matches max offset, then we do not need to lookup the metadata as we already have it. Really this is the bug fix. Without this, then we will proceed to read from the log segment, which results in an error like the following:

```
java.util.concurrent.ExecutionException: org.apache.kafka.common.KafkaException: java.io.EOFException: Failed to read `log header` from file channel `sun.nio.ch.FileChannelImpl@cc86429`. Expected to read 17 bytes, but reached end of file after reading 0 bytes. Started read from position 85.

	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at kafka.log.LogConcurrencyTest.testUncommittedDataNotConsumed(LogConcurrencyTest.scala:75)
	at kafka.log.LogConcurrencyTest.testUncommittedDataNotConsumedFrequentSegmentRolls(LogConcurrencyTest.scala:61)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: org.apache.kafka.common.KafkaException: java.io.EOFException: Failed to read `log header` from file channel `sun.nio.ch.FileChannelImpl@cc86429`. Expected to read 17 bytes, but reached end of file after reading 0 bytes. Started read from position 85.
	at org.apache.kafka.common.record.RecordBatchIterator.makeNext(RecordBatchIterator.java:40)
	at org.apache.kafka.common.record.RecordBatchIterator.makeNext(RecordBatchIterator.java:24)
	at org.apache.kafka.common.utils.AbstractIterator.maybeComputeNext(AbstractIterator.java:79)
	at org.apache.kafka.common.utils.AbstractIterator.hasNext(AbstractIterator.java:45)
	at org.apache.kafka.common.record.FileRecords.searchForOffsetWithSize(FileRecords.java:302)
	at kafka.log.LogSegment.translateOffset(LogSegment.scala:275)
	at kafka.log.LogSegment.read(LogSegment.scala:298)
	at kafka.log.Log.$anonfun$read$2(Log.scala:1515)
	at kafka.log.Log.read(Log.scala:2333)
	at kafka.log.LogConcurrencyTest$ConsumerTask.call(LogConcurrencyTest.scala:95)
	at kafka.log.LogConcurrencyTest$ConsumerTask.call(LogConcurrencyTest.scala:85)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.EOFException: Failed to read `log header` from file channel `sun.nio.ch.FileChannelImpl@cc86429`. Expected to read 17 bytes, but reached end of file after reading 0 bytes. Started read from position 85.
	at org.apache.kafka.common.utils.Utils.readFullyOrFail(Utils.java:966)
	at org.apache.kafka.common.record.FileLogInputStream.nextBatch(FileLogInputStream.java:68)
	at org.apache.kafka.common.record.FileLogInputStream.nextBatch(FileLogInputStream.java:41)
	at org.apache.kafka.common.record.RecordBatchIterator.makeNext(RecordBatchIterator.java:35)
	... 14 more
```

Reply:
Yeah I understand this is the bug fix, I'm only curious to see why we used 

```
emptyFetchDataInfo(maxOffsetMetadata, includeAbortedTxns)
```

not

```
emptyFetchDataInfo(startOffsetMetadata, includeAbortedTxns)
```

I realized it is just the same, so curious if you intentionally use `maxOffsetMetadata` for any other reasons, but I think it is just to avoid unnecessary `convertToOffsetMetadataOrThrow` :)

Reply:
It lets us skip the call to `convertToOffsetMetadataOrThrow` in order to find `startOffsetMetadata`. I tried changing the check below to `>=` and hit a similar error when trying to translate the offset in `LogSegment`. We might be able also be able to fix this by adding some additional checks in `LogSegment` to avoid the offset translation when the max position matches the start position of the requested offset.